keps/sig-cli/2906-kustomize-function-catalog/README.md

```diff
@@ -0,0 +1,667 @@
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Introduce a new API (`kind`) that will provide a mechanism to improve distribution and discovery of Kustomize plugins, for use with `Kustomization`, `Components`, and `Composition` resources.
+
+This new API will provide a standardized way to define a collection of one or more Kustomize plugins, as well as supporting KRM-style configuration resources, that can be consumed by Kustomize in order to automate the use of plugins and eliminate manual out-of-band discovery and installation steps, regardless of the packaging format. All Kustomize configuration objects (i.e. Kustomization, Component and Composition) will support plugin source configuration via this new kind. Ideally, we would like the new API to become a standard that other KRM-style transformer orchestrators such as KPT can adopt as well.
```

Comment:
I'm not sure what you mean by this question. Could you provide some more clarification about what this looks like, beyond the KRM function specification? The catalog is really intended to be a collection of KRM functions, so I'm having trouble understanding the goal of the abstraction for the single KRM function?

Reply:
I imagine we can define a `KRMFunction` CRD which is orchestrator neutral and is the smallest building blocking for other CRDs.

It may look like the following.  It contains a metadata of a function and its behavior (e.g. idempotent, network).

```
# krm-fn.yaml
apiVersion: config.k8s.io/v1alpha1
kind: KRMFunction
type: container|exec (container is default if omit)
container:
  image: <image-name>
  version: <semver>
  description: <1 or 2 sentences>
  sourceURL: <URL>
  exampleURLs:
    - example-url-1
    - example-url-2
  functionConfig:
    - apiVersion: example.com/v1
      kind: JavaApplication
      schemaURL: <http/file/OCI url>
    - ...
  idempotent: true|false
  requireNetwork: true|false
  requireVolumeMount: true|false
  maintainers:
    - foo@bar.com
  license: Apache 2.0
  tags:
    - mutator
exec:
  platforms:
    - bin: foo-amd64-linux
      os: linux
      arch: amd64
      uri: https://example.com/foo-amd64-linux.tar.gz
      sha256: <hash>
    - bin: foo-darwin-linux
      os: darwin
      arch: amd64
      uri: https://example.com/foo-darwin-linux.tar.gz
      sha256: <hash>
```

Then, we can potentially use a `corev1.List` to wrap multiple `KRMFunction` CRs. Or we can define a `KRMFunctionList`.

For example:
```
apiVersion: v1
kind: List
items:
  - apiVersion: config.k8s.io/v1alpha1
    kind: KRMFunction
    container:
      ...
  - apiVersion: config.k8s.io/v1alpha1
    kind: KRMFunction
    container:
      ...
```

The list is purely a collection of KRMFunction. It can be used to:
- Build a KRM function catalog website
- Consumed by a orchestrator or a IDE to enhance editing. (e.g. VSCode can suggest the available functions when a user editing a kustomize related file; auto-completion for `kustomize fn run --image=foo<TAB><TAB>`)

The KRMFunction CRD can also be used by the kustomize `Catalog` CRD:
```
apiVersion: kustomize.io/v1
kind: Catalog
metadata: 
  name: "example-co-plugins"
spec: 
  modules: 
  - apiVersion: example.com/v1
    kind: JavaApplication
    provider: 
      containerRef: krm-fn.yaml
  - apiVersion: example.com/v1
    kind: AnotherJavaApplication
    provider: 
      execRef: another-krm-fn.yaml
```

@jeremyrickard @KnVerey WDYT?